File: clients/ts/test/schema-ingest.test.mjs
import 'fake-indexeddb/auto';
import assert from 'node:assert/strict';
import test from 'node:test';
import { readFile, readdir } from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

import { OptoSyncClient, schema } from '../dist/index.js';

const { parseEnvelope, ingestEnvelope, IngestValidationError } = schema;

const fixturesDir = path.join(
  path.dirname(fileURLToPath(import.meta.url)),
  '../../../schema/fixtures',
);

test('accepts every shared valid fixture (cross-language corpus)', async () => {
  const dir = path.join(fixturesDir, 'valid');
  const files = (await readdir(dir)).filter((f) => f.endsWith('.json'));
  assert.ok(files.length >= 3);
  for (const file of files) {
    const raw = await readFile(path.join(dir, file), 'utf8');
    const envelope = await parseEnvelope(raw);
    assert.equal(envelope.formatVersion, 1, file);
  }
});

test('rejects every shared invalid fixture (cross-language corpus)', async () => {
  const dir = path.join(fixturesDir, 'invalid');
  const files = (await readdir(dir)).filter((f) => f.endsWith('.json'));
  assert.ok(files.length >= 4);
  for (const file of files) {
    const raw = await readFile(path.join(dir, file), 'utf8');
    await assert.rejects(
      parseEnvelope(raw),
      IngestValidationError,
      `expected ${file} to be rejected`,
    );
  }
});

test('accepts Blob-like inputs', async () => {
  const raw = await readFile(
    path.join(fixturesDir, 'valid/basic-upsert.json'),
    'utf8',
  );
  const blobLike = { text: async () => raw };
  const envelope = await parseEnvelope(blobLike);
  assert.equal(envelope.records.length, 1);
});

let databaseSequence = 0;
function makeClient() {
  databaseSequence += 1;
  return new OptoSyncClient({
    databaseName: `ingest-test-${databaseSequence}`,
    // Envelope payloads carry their own updatedAt; ingest must not restamp.
    stampUpdatedAt: false,
  });
}

test('ingestEnvelope queues one mutation per record, in file order', async () => {
  const client = makeClient();
  const raw = await readFile(
    path.join(fixturesDir, 'valid/nested-keyed-arrays.json'),
    'utf8',
  );
  const { receipts } = await ingestEnvelope(client, raw, { optimism: 'background' });
  assert.equal(receipts.length, 2);

  const pending = await client.pendingMutations();
  assert.equal(pending.length, 2);
  assert.equal(pending[0].operation, 'upsert');
  assert.equal(pending[0].recordId, 'doc-9');
  assert.equal(pending[0].baseRevision, '41');
  const payload = JSON.parse(pending[0].jsonPayload);
  assert.equal(payload.sections.length, 2);
  assert.equal(pending[1].operation, 'delete');
  assert.equal(pending[1].recordId, 'doc-10');
});

test('nothing queues when any record is invalid', async () => {
  const client = makeClient();
  const raw = await readFile(
    path.join(fixturesDir, 'invalid/missing-updated-at.json'),
    'utf8',
  );
  await assert.rejects(ingestEnvelope(client, raw), IngestValidationError);
  assert.equal((await client.pendingMutations()).length, 0);
});

File: clients/dart/test/schema_ingest_test.dart
import 'dart:io';

import 'package:drift/native.dart';
import 'package:opto_sync_client/opto_sync_client.dart';
import 'package:opto_sync_client/rx.dart' as rx;
import 'package:opto_sync_client/schema.dart' as schema;
import 'package:test/test.dart';

String locateCoreLibrary() {
  final env = Platform.environment['SYNCER_LIB_PATH'];
  if (env != null && env.isNotEmpty) return env;
  final starts = <String>[
    Directory.current.absolute.path,
    File.fromUri(Platform.script).parent.absolute.path,
  ];
  for (final start in starts) {
    var dir = Directory(start);
    for (var i = 0; i < 10; i++) {
      final buildDir =
          '${dir.path}${Platform.pathSeparator}syncer.c${Platform.pathSeparator}core${Platform.pathSeparator}build';
      if (Directory(buildDir).existsSync()) {
        final path = resolveSyncerLibraryPath(directory: buildDir);
        if (File(path).existsSync()) return path;
      }
      final parent = dir.parent;
      if (parent.path == dir.path) break;
      dir = parent;
    }
  }
  throw StateError('could not locate the syncer core library');
}

Directory locateFixtures() {
  var dir = Directory.current.absolute;
  for (var i = 0; i < 10; i++) {
    final candidate = Directory(
      '${dir.path}${Platform.pathSeparator}schema${Platform.pathSeparator}fixtures',
    );
    if (candidate.existsSync()) return candidate;
    final parent = dir.parent;
    if (parent.path == dir.path) break;
    dir = parent;
  }
  throw StateError('could not locate schema/fixtures above ${Directory.current.path}');
}

void main() {
  final fixtures = locateFixtures();

  List<File> jsonFiles(String subdir) =>
      Directory('${fixtures.path}${Platform.pathSeparator}$subdir')
          .listSync()
          .whereType<File>()
          .where((f) => f.path.endsWith('.json'))
          .toList()
        ..sort((a, b) => a.path.compareTo(b.path));

  test('accepts every shared valid fixture (cross-language corpus)', () {
    final files = jsonFiles('valid');
    expect(files.length, greaterThanOrEqualTo(3));
    for (final file in files) {
      final envelope = schema.parseEnvelope(file.readAsStringSync());
      expect(envelope.records, isNotEmpty, reason: file.path);
    }
  });

  test('rejects every shared invalid fixture (cross-language corpus)', () {
    final files = jsonFiles('invalid');
    expect(files.length, greaterThanOrEqualTo(4));
    for (final file in files) {
      expect(
        () => schema.parseEnvelope(file.readAsStringSync()),
        throwsA(isA<schema.IngestValidationException>()),
        reason: file.path,
      );
    }
  });

  test('ingestEnvelope queues one mutation per record, in file order', () async {
    final syncer = FfiSyncer(libraryPath: locateCoreLibrary());
    final db = OptoSyncDatabase(NativeDatabase.memory());
    addTearDown(db.close);
    final client = OptoSyncClient(db: db, syncer: syncer, stampUpdatedAt: false);

    final raw = File(
      '${fixtures.path}${Platform.pathSeparator}valid${Platform.pathSeparator}nested-keyed-arrays.json',
    ).readAsStringSync();
    final receipts = await schema.ingestEnvelope(
      client,
      raw,
      optimism: rx.Optimism.background,
    );
    expect(receipts, hasLength(2));

    final pending = await client.pendingMutations();
    expect(pending, hasLength(2));
    expect(pending[0].recordId, 'doc-9');
    expect(pending[0].baseRevision, '41');
    expect(pending[1].operation, 'delete');
    expect(pending[1].recordId, 'doc-10');
  });

  test('nothing queues when any record is invalid', () async {
    final syncer = FfiSyncer(libraryPath: locateCoreLibrary());
    final db = OptoSyncDatabase(NativeDatabase.memory());
    addTearDown(db.close);
    final client = OptoSyncClient(db: db, syncer: syncer);

    final raw = File(
      '${fixtures.path}${Platform.pathSeparator}invalid${Platform.pathSeparator}missing-updated-at.json',
    ).readAsStringSync();
    await expectLater(
      schema.ingestEnvelope(client, raw),
      throwsA(isA<schema.IngestValidationException>()),
    );
    expect(await client.pendingMutations(), isEmpty);
  });
}
